packages/financial-templates-lib/src/price-feed/BasisPriceFeed.js

```diff
@@ -0,0 +1,148 @@
+    }
+
+    this.spotPriceFeeds = spotPriceFeeds;
+    this.futurePriceFeeds = futurePriceFeeds;
+
+    this.lowerBound = lowerBound;
+    this.upperBound = upperBound;
+
+    // For convenience, concatenate all constituent price feeds.
+    this.allPriceFeeds = this.spotPriceFeeds.concat(this.futurePriceFeeds);
+
+    // Helper modules.
+    this.web3 = web3;
+    this.toBN = this.web3.utils.toBN;
+    this.toWei = this.web3.utils.toWei;
+    this.logger = logger;
+
+    // The precision that the user wants to return prices in must match all basket constituent price feeds and the denominator.
+    this.decimals = this.allPriceFeeds[0].getPriceFeedDecimals();
+
+    // Scale `number` by 10**decimals.
+    this.convertPriceFeedDecimals = number => {
+      // Converts price result to wei
+      // returns price conversion to correct decimals as a big number
+      return this.toBN(parseFixed(number.toString(), this.decimals).toString());
+    };
+  }
+
+  // Given lists of future and spot prices
+  // return the spread price, which is:
+  // 100 ((avg(future) - avg(spot))/(avg(spot)) + 1)
+  _getSpreadFromBasketPrices(futurePrices, spotPrices) {
+    // Compute future basket mean.
+    if (futurePrices.length === 0 || futurePrices.some(element => element === undefined || element === null)) {
+      return null;
+    }
+    const futureMean = this._computeMean(futurePrices);
+
+    // Second, compute the average of the spot pricefeeds.
+    if (spotPrices.length === 0 || spotPrices.some(element => element === undefined || element === null)) {
+      return null;
+    }
+    const spotMean = this._computeMean(spotPrices);
+
+    if (!spotMean || !futureMean) return null;
+
+    if (spotMean.eq(this.toBN("0"))) return this.convertPriceFeedDecimals("100");
+
+    let spreadValue = ((futureMean.sub(spotMean) / spotMean + 1) * 100).toFixed(this.decimals);
+
+    spreadValue = this.convertPriceFeedDecimals(spreadValue);
+
+    // Min + Max for clamping
+    let lowerBound = this.convertPriceFeedDecimals(this.lowerBound);
+    let upperBound = this.convertPriceFeedDecimals(this.upperBound);
+
+    if (spreadValue.lt(lowerBound)) {
+      spreadValue = lowerBound;
+    } else if (spreadValue.gt(upperBound)) {
+      spreadValue = upperBound;
+    }
+    return spreadValue;
+  }
+
+  getCurrentPrice() {
+    const futurePrices = this.futurePriceFeeds.map(priceFeed => priceFeed.getCurrentPrice());
+    const spotPrices = this.spotPriceFeeds.map(priceFeed => priceFeed.getCurrentPrice());
+
+    return this._getSpreadFromBasketPrices(futurePrices, spotPrices);
+  }
+
+  getHistoricalPrice(time) {
+    const futurePrices = this.futurePriceFeeds.map(priceFeed => priceFeed.getHistoricalPrice(time));
+    const spotPrices = this.spotPriceFeeds.map(priceFeed => priceFeed.getHistoricalPrice(time));
+
+    return this._getSpreadFromBasketPrices(futurePrices, spotPrices);
+  }
```

Comment:
We just made a change on master changing this function to async, so you'll need to make some minor changes here.

```suggestion
  async getHistoricalPrice(time) {
    const errors = [];
    const futurePrices = await Promise.all(this.futurePriceFeeds.map(priceFeed => priceFeed.getHistoricalPrice(time).catch(err => errors.push(err))));
    const spotPrices = await Promise.all(this.spotPriceFeeds.map(priceFeed => priceFeed.getHistoricalPrice(time).catch(err => errors.push(err))));
    
    if (errors.length > 0) {
      throw errors;
    }
    
    return this._getSpreadFromBasketPrices(futurePrices, spotPrices);
  }
```